### ISSUE TITLE
Kobolds can claim ennemy rooms/traps

### ISSUE THREAD

**Issue by hwoarangmy** [OpenDungeons/OpenDungeons#165]
See comments in #162 
An easy fix would be to check if the tile has a room/trap before trying to claim it. But it would avoid kobolds to go there if it is a trap.
So, kobolds should not consider claiming rooms but they should test if a trap is on a tile only after they moved on it.

**Comment by Bertram25** [comment]
Sounds fine to me. :)

**Comment by hwoarangmy** [comment]
Concerning claiming trap, should we use the same as in DK ? ie: when a tile is claimed the corresponding trap is destroyed ? If I remember well, it was @Danimal696 opinion to do that (I remember he said that in DK2, it was too easy to kill traps by attacking). I think we all agreed but I am not sure. If someone remember, please help my failing memory :)

**Comment by Bertram25** [comment]
My opinion is I'd rather say that until the trap is destroyed, one cannot claim the tile. Same for rooms.
But there might be flaws in this as well.

**Comment by hwoarangmy** [comment]
That brings us to how to destroy a trap ? At first, it should be hidden. If a player triggers the trap, it will become visible. Then, can creatures attack it ? If there are no creature to attack, during how much time will the trap become visible ?

**Comment by Bertram25** [comment]
You're right, certain traps should be hidden, but I think cannons shouldn't, so this should remain an option, IMHO.

I also do think the trap should remain visible once discovered as it won't move anyway, and it would look weird to let creatures rediscover it again and again.
Once discovered, then yes, I'd be for letting creatures destroy it before being able to claim the tile.
Thus, traps should have custom HP (not just the default 10 of a tile).
Feel free to correct me if you see something. :)

**Comment by hwoarangmy** [comment]
It's not a correction, just another discussion about the subject. I was able to find what I was talking about :
#139

**Comment by Danimal696** [comment]
My opinion is not mandatory, but im in favor of traps not being directly attacked even when discovered

**Comment by Bertram25** [comment]
If you don't want them attacked by fighters when discovered, then what should happen, in your own opinion?

**Comment by Danimal696** [comment]
they have to be triggered a number of times before becoming deactivated or claimed fast enougth so it doesnt trigger again when your kobolds are still over it (and die from it)

**Comment by Bertram25** [comment]
I'm fine for this with spikes, but this doesn't fit well for cannons, right?

**Comment by akien-mga** [comment]
I guess turrets and traps can't be handled exactly the same way.

**Comment by hwoarangmy** [comment]
> I guess turrets and traps can't be handled exactly the same way.

As @akien-mga suggested, we could separate traps and turrets. What do you think ?

**Comment by Bertram25** [comment]
Good idea. :)

**Comment by Danimal696** [comment]
yes, some traps like cannon should be destructible

**Comment by Bertram25** [comment]
Ok, then we all are ok that:
- If a trap is visible per default, it has to be destructed by fighters before letting workers claim the underlying tile. (Room-like.)
- If a trap is invisible, then workers have to claim the tile before being killed by the trap. Either by triggering it so much it is defunct, or by being quick enough.

Now, let's try to take a more complete example (without saying it will be actual future game content):
Visible traps:
- Cannons
- Bunker and such
- brambles, barbed, ...
- Doors

Invisible traps:
- Spikes
- Poison
- Acid
- Fire

Would those happen in game, should they be separated into two tabs or not? What would be their tabs names if they are separated?

I we can answer those last questions, we will be able to get specs out of those. Feel free to complete/correct me. :)

**Comment by Danimal696** [comment]
i dont think they should be separated in different tabs, perhaps the upper row of the trap tab goes to visible, and the lower row to invisible

**Comment by Bertram25** [comment]
> i dont think they should be separated in different tabs, perhaps the upper row of the trap tab goes to visible, and the lower row to invisible

It's an idea. Traps icons would be smaller but that's about it.

Thus, internally, we could either have two separate classes, or a mVisibleForCreatures boolean triggering either of the two described behaviour for traps.

Code speaking, what would be your favourite road, guys?

**Comment by hwoarangmy** [comment]
IMHO, a flag will be enough because I believe they will do nearly the same. We could also have turrets that are hidden and pops up when firing for a bit (while they can be attacked). Then, if still alive and no creature nearly, they could hide again.
But in any case, we have to do something to make sure players will know if a trap is to be attacked or to be claimed.

**Comment by Bertram25** [comment]
> We could also have turrets that are hidden and pops up when firing for a bit (while they can be attacked). Then, if still alive and no creature nearly, they could hide again.

IMHO, the creatures should know where the trap is, once it has popped up the first time. The trap would then be invisible once for this case.

> IMHO, a flag will be enough because I believe they will do nearly the same.

Fine to me.

**Comment by hwoarangmy** [comment]
> IMHO, the creatures should know where the trap is, once it has popped up the first time. The trap would then be invisible once for this case.

The problem is the one @Danimal696 said : once they see the trap, it will be very easy when there is FPP to kill the trap with 1 ranged unit

**Comment by Bertram25** [comment]
> The problem is the one @Danimal696 said : once they see the trap, it will be very easy when there is FPP to kill the trap with 1 ranged unit

Well, what is the point of the FPP mode if not doing exactly such kind of things? (Apart from enjoying dark corridors ;])
IMHO, it's not bad, it's even strategical. To me, the time you "loose" being a creature isn't spent managing the rest.

@Danimal696 Is there something wrong with my point of view?

**Comment by Danimal696** [comment]
For active traps like the cannon that is valid, posses some warlock and blast the cannon in pieces while dodging the bullets; but for floor traps it render them pointless, hence why i think they should be undestructible by force; its fine if they remain visible once triggered, so you can avoid it, but it wont go anywhere until you disarm it.

**Comment by Bertram25** [comment]
Ah I see. But I think we agree since the beginning. Let me split all the examples:
- The cannon, always visible. Fighters have to destroy it before proceeding.
- Spike traps are always invisible (even if visually shown when triggered) and only workers can disable them by claiming the tile or die in the attempt.
- hwoarangmy's mighty cannon that burst out of the ground and fires. It is invisible at start but once a creature has seen it, it is visible for the team and thus considered as a standard cannon.

That's what I had in mind. Are we in sync?

**Comment by Danimal696** [comment]
Spike traps can become visible when triggered once, since you cant forcefully destroy them, at least you know where they are, the lazy explanation could be the trap cant be destroyed since the mechanisn its fully underground

**Comment by Bertram25** [comment]
Then, there are several notions:
The trap is:
- visible for the player's team: Once discovered and visually shown.
- visible for the creature: Making fighters able to fight it, workers will try to claim if invisible.

So if we take the examples:
- The cannon example is visible for creatures and the player's team.
- Spikes are visible to none. Then, once triggered, they are visible for the player's team only.
- The super cannon is visible to none. Then, once triggered it becomes visible to both.

Is this more or less what you have in mind?

**Comment by Danimal696** [comment]
> visible for the creature: Making fighters able to fight it, workers will try to claim if invisible.

Fitghers should only engage traps like the cannon. Spike, poison and the like shouldnt be destructible by force, just by triggering a number of times or claiming the tile its on.

**Comment by Bertram25** [comment]
> Fitghers should only engage traps like the cannon, spike, poison and the like shouldnt be destructible by force, just by triggering a number of times or claiming the tile its on.

I think we agree, then. :)

**Comment by Danimal696** [comment]
> The super cannon is visible to none. Then, once triggered it becomes visible to both.

Uhmm, i would call it the "camo cannon" and make it a step above the normal cannon, same damage but hiding capabilities, it shoots and is made invisible again when no enemies are near.

**Comment by Bertram25** [comment]
Don't worry, those were just examples. ;) But this helps when making specs.

**Comment by hwoarangmy** [comment]
Ok, let's start with those 3 types. At first, we can just show/hide "camo cannon" or play with its size to do as there was an animation. Later, it would be nice to have a special animation for this kind.

**Comment by Bertram25** [comment]
Ok :)

**Comment by hwoarangmy** [comment]
The room part is fixed in #169 
As decided, claiming trap is normal so it is not a bug. In my next PR, claiming trap will destroy them (even canons). If after testing, we want to change, we will open another issue.
I close this one
